### PR TITLE
Add dynamically setup containerPorts in proxy infrastructure

### PR DESCRIPTION
**Pull request by ShawnHXH**
Add dynamically setup containerPorts in proxy infrastructure

refer to #1032

### CODE REVIEW

File: internal/infrastructure/kubernetes/proxy_deployment.go [hunk 1]
```diff
@@ -140,6 +140,22 @@ func expectedProxyContainers(infra *ir.Infra) ([]corev1.Container, error) {
 		},
 	}
 
+	for _, listener := range infra.Proxy.Listeners {
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/proxy_deployment.go:143]: dont you need to delete the port config on L130-140 ?

File: internal/infrastructure/kubernetes/proxy_deployment.go [hunk 2]
```diff
@@ -140,6 +140,22 @@ func expectedProxyContainers(infra *ir.Infra) ([]corev1.Container, error) {
...
+		for _, lp := range listener.Ports {
+			if corev1.Protocol(lp.Protocol) == corev1.ProtocolTCP {
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/proxy_deployment.go:145]: I dont think this is needed

File: internal/infrastructure/kubernetes/proxy_deployment.go [hunk 3]
```diff
@@ -140,6 +140,22 @@ func expectedProxyContainers(infra *ir.Infra) ([]corev1.Container, error) {
...
+				if lp.ContainerPort == envoyHTTPPort || lp.ContainerPort == envoyHTTPSPort {
+					continue
+				}
+			}
+			port := corev1.ContainerPort{
+				Name:          lp.Name,
+				ContainerPort: lp.ContainerPort,
+				Protocol:      corev1.Protocol(lp.Protocol),
+			}
+			ports = append(ports, port)
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/proxy_deployment.go:155]: can you also add some test cases in proxy_deployment_test.go ?

File: internal/infrastructure/kubernetes/proxy_deployment.go (hunk 1 above)
Comment (ShawnHXH) [reply internal/infrastructure/kubernetes/proxy_deployment.go:143]: sure

File: internal/infrastructure/kubernetes/proxy_deployment.go (hunk 2 above)
Comment (ShawnHXH) [reply internal/infrastructure/kubernetes/proxy_deployment.go:145]: fixed

File: internal/infrastructure/kubernetes/proxy_deployment.go (hunk 3 above)
Comment (ShawnHXH) [reply internal/infrastructure/kubernetes/proxy_deployment.go:138]: done

File: internal/infrastructure/kubernetes/proxy_deployment_test.go [hunk 4]
```diff
@@ -127,6 +127,31 @@ func TestExpectedProxyDeployment(t *testing.T) {
 
 	infra.Proxy.GetProxyMetadata().Labels[gatewayapi.OwningGatewayNamespaceLabel] = "default"
 	infra.Proxy.GetProxyMetadata().Labels[gatewayapi.OwningGatewayNameLabel] = infra.Proxy.Name
+	infra.Proxy.Listeners = []ir.ProxyListener{
```
[the remark this answers is not in the document]
Comment (ShawnHXH) [reply internal/infrastructure/kubernetes/proxy_deployment_test.go:130]: Sure, the separate test has been added.

**Review by arkodg — APPROVED** [review]
LGTM, thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**